lib/Doctrine/DBAL/SQLParserUtils.php

```diff
@@ -73,6 +73,42 @@ static public function getPlaceholderPositions($statement, $isPositional = true)
         return $paramMap;
     }
 
+    static private function startsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        return (substr($haystack, 0, $length) === $needle);
+    }
+
+    static private function endsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        if ($length == 0) {
+            return true;
+        }
+
+        return (substr($haystack, -$length) === $needle);
+    }
+
+    static private function isCollectionType($type)
+    {
+        if ($type === Connection::PARAM_INT_ARRAY || $type === Connection::PARAM_STR_ARRAY)
+            return true;
+
+        return static::startsWith($type, '[') && static::endsWith($type, ']');
+    }
+
+    static private function removeCollectionMarker($type)
+    {
+        if (!static::isCollectionType($type))
+            throw SQLParserUtilsException::notCollectionType($type);
+        
+        if (is_numeric($type))
+            return $type - Connection::ARRAY_PARAM_OFFSET;
+
+        // [my_custom_type] -> my_custom_type
```

Comment:
`"[foo]"` being considered `"special scenario in which the type represents an array of foo"` is not viable. What you are doing here is moving away from a simple map type, and converting type discovery into a function that actually creates more complex types.

I'd rather not have the feature at all than have it as magic strings.

 > We're already using this on our own project, and this is working well :)

The fact that it is working has nothing to do with correctness, sorry :-\